File: python/8_function2qtt.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Conversion of one-dimensional function into QTT representation
"""

import matplotlib.pyplot as plt
import numpy as np


def main():
    depth = 8
    npoints = 2**depth
    cutoff = 1e-10
    max_rank = 4

    """target function"""
    x = np.linspace(0.0, 1.0, npoints)
    y = np.exp(x)
    # y = np.sin(x)
    # y = np.exp(-(((x - 0.5) / 0.1) ** 2)) / (0.1 * np.sqrt(np.pi))
    if depth < 5:
        print(f"x = {x}")
        print(f"y = {y}\n")

    """QTT decomposition"""
    yt = y.copy()
    qtt = []
    rank = 1
    for k in range(0, depth - 1):
        print(f"depth: {k}")
        yt = yt.reshape((rank * 2, -1))
        U, S, Vt = np.linalg.svd(yt, full_matrices=False)
        print(f"singular values: {S}")
        S = S[S > cutoff * S[0]]
        S = S[: min(len(S), max_rank)]
        rank_new = len(S)
        U = U[:, :rank_new]
        Vt = Vt[:rank_new, :]
        if k > 0:
            U = U.reshape((rank, 2, rank_new))
        print(f"tensor shape: {U.shape}\n")
        qtt.append(U)
        yt = np.dot(np.diag(S), Vt)
        rank = rank_new
    yt = yt.reshape((rank, 2))
    print(f"depth: {depth - 1}")
    print(f"tensor shape: {yt.shape}\n")
    qtt.append(yt)
    if depth < 5:
        print(f"qtt: {qtt}\n")

    yr = qtt[0]
    for k in range(1, depth - 1):
        yr = np.einsum("ij,jkl->ikl", yr, qtt[k])
        yr = yr.reshape(yr.shape[0] * yr.shape[1], yr.shape[2])
    yr = np.einsum("ij,jk->ik", yr, qtt[depth - 1])
    yr = yr.reshape(-1)
    print(f"error = {np.linalg.norm(y - yr) / npoints}\n")

    if depth < 5:
        print(f"reconstructed y = {yr}\n")
    plt.figure()
    plt.plot(x, y, label="target")
    if npoints <= 32:
        plt.plot(x, yr, "o", label="QTT")
    else:
        plt.plot(x[:: npoints // 32], yr[:: npoints // 32], "o", label="QTT")
    plt.legend()
    plt.show()


if __name__ == "__main__":
    main()

File: julia/8_function2qtt.jl
#!/usr/bin/env julia
"""
Conversion of one-dimensional function into QTT representation (Julia)
"""

using LinearAlgebra
using TensorOperations
using Plots

function main()
    depth = 4
    npoints = 2^depth
    cutoff = 1e-10
    max_rank = 4

    # target function
    x = range(0.0, 1.0; length=npoints)
    y = exp.(x)
    # y = cos.(x)
    # y = @. exp(-(((x - 0.5) / 0.1)^2)) / (0.1 * sqrt(pi))

    if depth < 5
        println("x = ", collect(x))
        println("y = ", y, "\n")
    end

    # QTT decomposition
    yt = copy(y)
    qtt = Array{Float64}[]
    rank = 1
    for k in 0:(depth-2)
        println("depth: $k")
        yt = reshape(permutedims(reshape(yt, rank, :, 2), (3, 1, 2)), 2 * rank, :)
        F = svd(yt; full=false)
        S = F.S
        println("singular values: ", S)
        rank_new = findall(>(cutoff * S[1]), S)
        rank_new = min(length(rank_new), max_rank)
        U = F.U[:, 1:rank_new]
        Vt = F.Vt[1:rank_new, :]
        if k > 0
            U = permutedims(reshape(U, 2, rank, rank_new), (2, 1, 3))
        end
        push!(qtt, U)
        yt = (Diagonal(S[1:rank_new]) * Vt)
        rank = rank_new
    end
    println("depth: $(depth-1)")
    push!(qtt, yt)
    if depth < 5
        println("qtt: ", qtt, "\n")
    end

    # reconstruction
    yr = qtt[1]::Matrix{Float64}
    for k in 2:(depth-1)
        # not [i, k, l] but [k, i, l] because of column-major order
        @tensor tmp[k, i, l] := yr[i, j] * qtt[k][j, k, l]
        yr = reshape(tmp, size(tmp, 2) * size(tmp, 1), size(tmp, 3))
    end
    @tensor yr2[i, k] := yr[i, j] * qtt[end][j, k]
    state = vec(transpose(yr2))
    err = norm(y .- state) / npoints
    println("error = $err\n")

    if depth < 5
        println("reconstructed y = ", state, "\n")
    end
    plt = plot(x, y, label="target")
    if npoints <= 32
        plot!(x, state, seriestype=:scatter, label="QTT")
    else
        step = Int(cld(npoints, 32))
        plot!(x[1:step:end], state[1:step:end], seriestype=:scatter, label="QTT")
    end
    display(plt)

    plt = plot(x, y - state, label="error")
    display(plt)
end

main()
